File: Ant/chapter3/python/face_the_right_way.py
def ns(f):
    return next(f).strip()


with open("../testset/face_the_right_way/test1.txt", 'r') as f:
    N = int(ns(f))
    d = list(map(lambda x: 1 if x == 'B' else 0, ns(f)))


def calc(k):
    f = [0] * N
    s = 0
    res = 0
    for i in range(N - k + 1):
        if (d[i] + s) % 2 != 0:
            f[i] = 1
            res += 1
        s += f[i]
        if i - k + 1 >= 0:
            s -= f[i - k + 1]
    for i in range(N - k + 1, N):
        if (d[i] + s) % 2 != 0:
            return -1
        if i - k + 1 >= 0:
            s -= f[i - K + 1]
    return res


M, K = N, 1
for k in range(1, N + 1):
    m = calc(k)
    if 0 <= m < M:
        M = m
        K = k
print(M, K)

File: Ant/chapter3/cpp/face_the_right_way.cpp
#include <fstream>
#include <memory.h>
#include <stdio.h>

#define MAX_N 5000

using namespace std;

int N, M, K;
int dir[MAX_N]; // 0 := F, 1 := B
int f[MAX_N];

int calc(int k) {
  memset(f, 0, sizeof(f));
  int res = 0;
  int s = 0;
  for (int i = 0; i + k <= N; ++i) {
    if ((dir[i] + s) % 2 != 0) {
      ++res;
      f[i] = 1;
    }
    s += f[i];
    if (i - k + 1 >= 0) {
      s -= f[i - k + 1];
    }
  }
  for (int i = N - k + 1; i < N; ++i) {
    if ((dir[i] + s) % 2 != 0) {
      return -1;
    }
    if (i - k + 1 >= 0) {
      s -= f[i - k + 1];
    }
  }
  return res;
}

int main() {
  ifstream ifs("../testset/face_the_right_way/test1.txt");
  ifs >> N;
  for (int i = 0; i < N; ++i) {
    char d;
    ifs >> d;
    dir[i] = d == 'B';
  }
  K = 1;
  M = N;
  for (int k = 1; k <= N; ++k) {
    int m = calc(k);
    if (0 <= m && m < M) {
      M = m;
      K = k;
    }
  }
  printf("%d %d\n", M, K);
}